subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/IncrementalAnnotationProcessingTest.groovy

```diff
@@ -0,0 +1,437 @@
+        given:
+        appWithIncapProcessorOnly()
+
+        when:
+        succeeds "compileJava"
+
+        then:
+        getMappingFile().exists()
+    }
+
+    @Ignore
+    def "malformed mapping file results in full build on next compilation"() {
+        given:
+        appWithIncapProcessorOnly()
+        run "compileJava"
+
+        when:
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*trivial change*/ }"
+        getMappingFile().text = "<malformed contents>"
+
+        then:
+        // TODO:  Need to get this test working.  Somehow the mapping file, whose path doesn't change,
+        // is being rewritten (and repaired) between the line above and the "fails" line below.
+        fails "app:compileJava"
+
+        when:
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*trivial change 2*/ }"
+
+        then:
+        succeeds "app:compileJava"
+    }
+
+    def "malformed meta-inf file results in non-incremental build"() {
+        given:
+        appWithMalformedProcessorOnly()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*trivial change*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        // Verify that it was non-incremental (everything recompiled).
+        outputs.recompiledClasses 'App', 'B', 'AppIncremental'
+    }
+
+    def "delete generated file - Gradle recompiles the source file"() {
+        given:
+        appWithIncapProcessorOnly()
+        run 'compileJava'
+        def genFile = file('app/build/generated-sources/AppIncremental.java')
+
+        when:
+        Files.deleteIfExists(genFile.toPath())
+        succeeds 'app:compileJava'
+
+        then:
+        genFile.exists()
+    }
+
+    def "unrelated java file changed - generated sources are untouched"() {
+        given:
+        appWithIncapProcessorOnly()
+        run "compileJava"
+        def genfile = file('app/build/generated-sources/AppIncremental.java')
+        def timestamp = genfile.lastModified()
+
+        when:
+        file('app/src/main/java/B.java').text = "class B { /* unrelated change */ }"
+        succeeds 'app:compileJava'
+
+        then:
+        outputs.recompiledClasses 'App', 'B', 'AppIncremental'
+        timestamp == genfile.lastModified()
+    }
+
+    def "incremental processor correctly updates generated file when source changes"() {
+        given:
+        incrementalAppWithMethodsOnly()
+        run "compileJava"
+        def genfile = file('app/build/generated-sources/AppIncremental.java')
+        def srcfile = file('app/src/main/java/App.java')
+
+        when:
+        !genfile.text.contains("public void generatedBar()")
+        srcfile.text = """
+            @Incremental class App {
+               public void foo() {}
+               public void bar() {}
+            }
+        """
+        succeeds 'app:compileJava'
```

Comment:
I don't understand the structure of the test, and a few others. It seems to me that assertions are performed during the `when` phase: `succeeds compileJava`, or `!genfile.text.contains("public void generatedBar()")`. These should go in the `then` phase / part of the test.